cclunit-framework-source/src/test/ccl/ut_cclut_mock.inc

```diff
@@ -0,0 +1,409 @@
+        1.0)
+    call cclutAssertf8Equal(CURREF, "test_cclutExecuteProgramWithMocks_happy",
+        executeReply->regular_join[1].encounter_id, 4.0)
+    call cclutAssertvcEqual(CURREF, "test_cclutExecuteProgramWithMocks_happy",
+        executeReply->regular_join[1].encounter_alias, "test alias")
+
+    call cclutAsserti4Equal(CURREF, "test_cclutExecuteProgramWithMocks_happy", size(executeReply->left_join, 5), 3)
+    call cclutAssertf8Equal(CURREF, "test_cclutExecuteProgramWithMocks_happy", executeReply->left_join[1].person_id,
+        1.0)
+    call cclutAssertf8Equal(CURREF, "test_cclutExecuteProgramWithMocks_happy",
+        executeReply->left_join[1].encounter_id, 4.0)
+    call cclutAssertvcEqual(CURREF, "test_cclutExecuteProgramWithMocks_happy",
+        executeReply->left_join[1].encounter_alias, "test alias")
+    call cclutAssertf8Equal(CURREF, "test_cclutExecuteProgramWithMocks_happy", executeReply->left_join[2].person_id,
+        2.0)
+    call cclutAssertf8Equal(CURREF, "test_cclutExecuteProgramWithMocks_happy",
+        executeReply->left_join[2].encounter_id, 5.0)
+    call cclutAssertvcEqual(CURREF, "test_cclutExecuteProgramWithMocks_happy",
+        executeReply->left_join[2].encounter_alias, "")
+    call cclutAssertf8Equal(CURREF, "test_cclutExecuteProgramWithMocks_happy", executeReply->left_join[3].person_id,
+        3.0)
+    call cclutAssertf8Equal(CURREF, "test_cclutExecuteProgramWithMocks_happy",
+        executeReply->left_join[3].encounter_id, 0.0)
+    call cclutAssertvcEqual(CURREF, "test_cclutExecuteProgramWithMocks_happy",
+        executeReply->left_join[3].encounter_alias, "")
+
+    set stat = alterlist(executeReply->rdb_join, 1)
+    free define rtl2
+    define rtl2 is "ccluserdir:cclut_happy.dat"
+    select into "nl:"
+    from rtl2t r
+    head report
+        executeReply->rdb_join[1].columns = r.line
+    foot report
+        executeReply->rdb_join[1].data = r.line
+    with nocounter
+
+    call cclutAssertVcOperator(CURREF, "test_cclutExecuteProgramWithMocks_happy", executeReply->rdb_join[1].columns,
+        "regexplike", "PERSON_ID.*ENCOUNTER_ID.*PERSON_ID.*ENCOUNTER_ID.*ENCOUNTER_ALIAS")
+    call cclutAssertVcOperator(CURREF, "test_cclutExecuteProgramWithMocks_happy", executeReply->rdb_join[1].data,
+        "regexplike", "1.*4.*1.*4.*test alias")
+end ;test_cclutExecuteProgramWithMocks_happy
+
+declare CCLUT::InternalSubroutine(null) = null with protect
+
+subroutine (CCLUT::InternalSubroutine(null) = null with protect)
+    set executeReply->test_subroutine = 1
+end
+
+/* test_cclutExecuteProgramWithMocks_namespace **************************************************************
+*  Scenario: Executes the program using the supplied namespace                                              *
+************************************************************************************************************/
+subroutine test_cclutExecuteProgramWithMocks_namespace(null)
```

Comment:
Before I make this change, I want to make sure this is what you want because the test is already doing this with one small difference.

The executeReply structure has three items in it used by the subroutines - public_subroutine, test_subroutine, and mock_subroutine.

In ut_cclut_execute_mocks_happy, there's only public::internalSubroutine which sets the public_subroutine value to 1.

Just above this test, cclut::internalSubroutine is defined which sets the test_subroutine value to 1.  So since this test checks that public_subroutine is 0 and test_subroutine is 1, it's functionally similar to what you're suggesting.  The only difference is that cclut::internalSubroutine is defined in the test .inc instead of in the .prg.  If you want me to actually move it into the program, I can.

Reply:
I'm really confused by this, especially this statement:

> The gist of my original comment is that (1) is not tested.

It is tested exactly as you describe, so I don't know how to change it for what you're asking for.

> Demonstrate that when "NS" is passed as the third parameter to cclutExecuteProgramWithMocks then
SUT invokes subroutine NS::SUB defined by the test rather than PUBLIC::SUB defined by SUT when the SUT code invokes SUB (without specifying the namespace).

So bringing your example to the actual code, "NS" would be "CCLUT" and "SUB" is "InternalSubroutine" (now "internalSubroutine").  cclut_execute_mocks_happy is the SUT, and it defines "PUBLIC:InternalSubroutine" (now "PUBLIC:internalSubroutine") and calls "internalSubroutine" without the namespace.

I don't want to dwell on the old version though since I have made more changes based on your recommendations to streamline certain elements (e.g. having the subroutine defined within the test instead of outside of it).  However, the behavior of the test is still intact.

I'd like to narrow this recommendation down to the specific action items that you need me to do.  Since I still believe that 1) is being tested, here's what I'm thinking (please correct if necessary):

1. From you, I need specific instructions for what I need to change so that you feel point 1 is being tested.
2. For me, I'll add the record and variable testing to the namespace test.
3. For me, I'll create a separate .inc to test the carry-forward aspect (I already do a clean up of all mocks in the teardown currently for the regular ut_cclut_mock.inc).

Does that sound reasonable?

Reply:
You are correct. I misread things (failed to see cclutRemoveAllMocks in teardown).

Observe that the following code achieves the same goals and the _mock_subroutine test could be similarly simplified.
```
record executeReply (1 test_subroutine = i4)
call cclutAddMockImplementation("REPLY", "EXECUTEREPLY")
call cclutExecuteProgramWithMocks("cclut_execute_mocks_happy", "", "CCLUT")
call cclutAsserti4Equal(CURREF, "test_cclutExecuteProgramWithMocks_namespace", executeReply->test_subroutine, 1)
```

Reply:
I think you might still be looking at the old version of this because I moved the variables outside of the record structure to streamline it a bit (similar to what you're suggesting here).  Can you check the latest version here:
https://github.com/cerner/cclunit-framework/pull/7/files#diff-59424b610c97f031c4974a2164429109R331

And let me know if you still want me to make the changes you're suggesting?

Reply:
I was. 

The gist of my suggestion still holds. The test could skip declaring/checking variables not set by the test's mock implementation and skip passing parameters into the script. 

If the real or other mock implementation ended up being called, the test would still 'fail' because the requisite variables and/or parameters would not available, and errors would be generated.

Reply:
Alright, I'll take them out.